### PR TITLE
[#6369] Convert metadata attributeValues to jsonb

### PR DESCRIPTION
**Pull request by vietnguyen**

### CODE REVIEW

**Comment by jason-p-pickering** [comment]
Can one of the admins verify this patch?

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/CustomAttributeSerializer.java [hunk 1]
```diff
@@ -28,28 +28,35 @@
...
+    extends JsonSerializer<Attribute>
+{
     @Override
-    public Schema getSchema()
+    public void serialize( Attribute attribute, JsonGenerator jsonGenerator, SerializerProvider serializers )
+        throws IOException
     {
-        Schema schema = new Schema( AttributeValue.class, SINGULAR, PLURAL );
-        schema.setOrder( Ordered.HIGHEST_PRECEDENCE );
+        ToXmlGenerator xmlGenerator = null;
```
Comment (mortenoh) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/CustomAttributeSerializer.java]: add newline

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/CustomAttributeSerializer.java [hunk 2]
```diff
@@ -28,28 +28,35 @@
...
+        if ( jsonGenerator instanceof ToXmlGenerator )
+        {
+            xmlGenerator = (ToXmlGenerator) jsonGenerator;
+        }
 
-        return schema;
+        jsonGenerator.writeStartObject();
```
Comment (mortenoh) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/CustomAttributeSerializer.java]: add newline

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/CustomAttributeSerializer.java [hunk 3]
```diff
@@ -28,28 +28,35 @@
...
+        if ( xmlGenerator != null )
+        {
+            xmlGenerator.setNextIsAttribute( true );
+            xmlGenerator.setNextName( null );
+        }
```
Comment (mortenoh) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/CustomAttributeSerializer.java]: add newline

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/attribute/DefaultAttributeService.java [hunk 4]
```diff
@@ -174,64 +200,73 @@ public Attribute getAttributeByCode( String code )
...
             return;
         }
 
-        if ( attributeValue.getAttribute().isUnique() )
+        Attribute attribute = getAttribute( attributeValue.getAttribute().getUid() );
+
+        if ( Objects.isNull( attribute ) || !attribute.getSupportedClasses().contains( object.getClass() ) )
+        {
+            return;
+        }
```
Comment (mortenoh) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/attribute/DefaultAttributeService.java]: add newline

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/common/DefaultIdentifiableObjectManager.java [hunk 5]
```diff
@@ -544,6 +548,23 @@ public void delete( IdentifiableObject object, User user )
...
 
+    @Override
+    @Transactional( readOnly = true)
+    public <T extends IdentifiableObject> List getAllValuesByAttributes( Class<T> klass, List<Attribute> attributes )
+    {
+        List<T> objects = getAllByAttributes( klass, attributes );
+        Set<String> attributeIds = attributes.stream().map( attribute -> attribute.getUid() )
+            .collect( Collectors.toSet() );
+
+        ArrayList<String> result = new ArrayList<String>();
```
Comment (mortenoh) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/common/DefaultIdentifiableObjectManager.java]: List<String>

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/common/DefaultIdentifiableObjectManager.java [hunk 6]
```diff
@@ -544,6 +548,23 @@ public void delete( IdentifiableObject object, User user )
         return (List<T>) store.getAllByAttributes( attributes );
     }
 
+    @Override
+    @Transactional( readOnly = true)
+    public <T extends IdentifiableObject> List getAllValuesByAttributes( Class<T> klass, List<Attribute> attributes )
```
Comment (mortenoh) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/common/DefaultIdentifiableObjectManager.java]: Returning list without a type specified?

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/common/hibernate/HibernateIdentifiableObjectStore.java [hunk 7]
```diff
@@ -28,22 +28,24 @@
...
 import org.apache.commons.logging.Log;
 import org.apache.commons.logging.LogFactory;
 import org.hibernate.Criteria;
 import org.hibernate.SessionFactory;
-import org.hibernate.criterion.DetachedCriteria;
-import org.hibernate.criterion.Disjunction;
-import org.hibernate.criterion.Property;
-import org.hibernate.criterion.Restrictions;
-import org.hibernate.criterion.Subqueries;
+import org.hibernate.criterion.*;
```
Comment (mortenoh) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/common/hibernate/HibernateIdentifiableObjectStore.java]: Set a higher threshold for import collapse, we generally only do it if really required

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/CustomAttributeSerializer.java (hunk 2 above)
Comment (vietnguyen) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/CustomAttributeSerializer.java]: Fixed

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/attribute/DefaultAttributeService.java (hunk 4 above)
Comment (vietnguyen) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/attribute/DefaultAttributeService.java]: Fixed

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/common/DefaultIdentifiableObjectManager.java (hunk 5 above)
Comment (vietnguyen) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/common/DefaultIdentifiableObjectManager.java]: Fixed

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/common/hibernate/HibernateIdentifiableObjectStore.java (hunk 7 above)
Comment (vietnguyen) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/common/hibernate/HibernateIdentifiableObjectStore.java]: Fixed

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/common/DefaultIdentifiableObjectManager.java (hunk 6 above)
Comment (vietnguyen) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/common/DefaultIdentifiableObjectManager.java]: Fixed

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/CustomAttributeSerializer.java (hunk 3 above)
Comment (vietnguyen) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/CustomAttributeSerializer.java]: fixed

File: dhis-2/dhis-services/dhis-service-dxf2/src/test/java/org/hisp/dhis/dxf2/datavalueset/DataValueSetServiceTest.java [hunk 8]
```diff
@@ -362,7 +384,17 @@ public void testImportDataValueSetXmlPreheatCache()
...
-        CompleteDataSetRegistration registration = registrationService.getCompleteDataSetRegistration( dsA, peA, ouA, ocDef );
+        CompleteDataSetRegistration registration = null;
+        try
+        {
+
+            registration = registrationService.getCompleteDataSetRegistration( dsA, peA, ouA, ocDef );
+        }
+        catch ( Exception ex )
+        {
+            System.out.println( DebugUtils.getStackTrace( ex ) );
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-dxf2/src/test/java/org/hisp/dhis/dxf2/datavalueset/DataValueSetServiceTest.java]: Remove system.out statement. Fix issue properly.

File: dhis-2/dhis-support/dhis-support-db-migration/src/main/resources/org/hisp/dhis/db/migration/2.33/V2_33_14__Add_jsonb_attributevalues_columns.sql [hunk 9]
```diff
@@ -0,0 +1,30 @@
+alter table dataelement add column if not exists attributevalues  jsonb  default '{}'::jsonb;
+alter table dataset add column if not exists  attributevalues jsonb  default '{}'::jsonb;
```
Comment (larshelge) [inline dhis-2/dhis-support/dhis-support-db-migration/src/main/resources/org/hisp/dhis/db/migration/2.33/V2_33_14__Add_jsonb_attributevalues_columns.sql]: One space too much after exists in all statements.

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/common/DefaultIdentifiableObjectManager.java [hunk 10]
```diff
@@ -544,6 +548,23 @@ public void delete( IdentifiableObject object, User user )
         return (List<T>) store.getAllByAttributes( attributes );
     }
 
+    @Override
+    @Transactional( readOnly = true)
+    public <T extends IdentifiableObject> List<String> getAllValuesByAttributes( Class<T> klass, List<Attribute> attributes )
+    {
+        List<T> objects = getAllByAttributes( klass, attributes );
+        Set<String> attributeIds = attributes.stream().map( attribute -> attribute.getUid() )
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/common/DefaultIdentifiableObjectManager.java]: Is there a way to make this more effective by utilizing a database query?

File: dhis-2/dhis-support/dhis-support-test/src/main/java/org/hisp/dhis/TransactionalIntegrationTestBase.java [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
```
Comment (larshelge) [inline dhis-2/dhis-support/dhis-support-test/src/main/java/org/hisp/dhis/TransactionalIntegrationTestBase.java]: Can we add some more unit tests for the attribute storage and query in general?

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/common/DefaultIdentifiableObjectManager.java (hunk 10 above)
Comment (vietnguyen) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/common/DefaultIdentifiableObjectManager.java]: Fixed by using double query

File: dhis-2/dhis-support/dhis-support-test/src/main/java/org/hisp/dhis/TransactionalIntegrationTestBase.java (hunk 11 above)
Comment (vietnguyen) [reply dhis-2/dhis-support/dhis-support-test/src/main/java/org/hisp/dhis/TransactionalIntegrationTestBase.java]: All attributeValues methods are tested in AttributeValueServiceTest.

File: dhis-2/dhis-services/dhis-service-dxf2/src/test/java/org/hisp/dhis/dxf2/datavalueset/DataValueSetServiceTest.java (hunk 8 above)
Comment (vietnguyen) [reply dhis-2/dhis-services/dhis-service-dxf2/src/test/java/org/hisp/dhis/dxf2/datavalueset/DataValueSetServiceTest.java]: Fixed

File: dhis-2/dhis-support/dhis-support-db-migration/src/main/resources/org/hisp/dhis/db/migration/2.33/V2_33_14__Add_jsonb_attributevalues_columns.sql (hunk 9 above)
Comment (vietnguyen) [reply dhis-2/dhis-support/dhis-support-db-migration/src/main/resources/org/hisp/dhis/db/migration/2.33/V2_33_14__Add_jsonb_attributevalues_columns.sql]: fixed

**Pull request merged**